src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -100,11 +103,35 @@ public async Task InitializedAsync()
         }
 
         [JsonRpcMethod(Methods.ShutdownName)]
-        public object? Shutdown(CancellationToken _) => null;
+        public Task ShutdownAsync(CancellationToken _)
+        {
+            Contract.ThrowIfTrue(_shuttingDown, "Shutdown has already been called.");
+
+            _shuttingDown = true;
+            _diagnosticService.DiagnosticsUpdated -= DiagnosticService_DiagnosticsUpdated;
+
+            return Task.CompletedTask;
+        }
 
         [JsonRpcMethod(Methods.ExitName)]
-        public void Exit()
+        public Task ExitAsync(CancellationToken _)
         {
+            Contract.ThrowIfFalse(_shuttingDown, "Shutdown has not been called yet.");
+
+            try
+            {
+                if (!_jsonRpc.IsDisposed)
+                {
+                    _jsonRpc.Dispose();
```

Comment:
Dispose's rule per the Framework rules: "If an object's Dispose method is called more than once, the object must ignore all calls after the first one." -- does StreamJsonRpc break that?

Reply:
Looks like it does judging by its implementation: https://github.com/microsoft/vs-streamjsonrpc/blob/e5fad2a29bbe9f8ca22da06121854c2f925634c5/src/StreamJsonRpc/JsonRpc.cs#L1190

Reply:
It's following the framework, right? The first check it does is whether it's already been disposed.

Reply:
"it does" as in it does ignore calls after the first one. I can remove